### ISSUE TITLE
Add dyamic array to P4

### ISSUE THREAD

**Issue by hesingh** [p4lang/p4-spec#1264]
This would help any packet buffering used by 5G cellular. Packet buffering is also used for reassembly in end nodes.

From: Mihai Budiu <[mbudiu@vmware.com](mailto:mbudiu@vmware.com)> 
Sent: Friday, June 18, 2021 2:17 AM
To: [hemant@mnkcg.com](mailto:hemant@mnkcg.com); Gergely Pongracz <[Gergely.Pongracz@ericsson.com](mailto:Gergely.Pongracz@ericsson.com)>; [jnfoster@cs.cornell.edu](mailto:jnfoster@cs.cornell.edu)
Cc: [p4-design@lists.p4.org](mailto:p4-design@lists.p4.org)
Subject: RE: [P4-design] Re: new language features for "5G" and "edge" use cases
 
Packet_in is an extern. There are no operations on extern except instantiation and method calls.
In particular, there is no assignment between externs (except compile-time binding as parameters).
If you want to do something like this you will probably have to invent a new extern to represent a dynamic array.
 
Mihai
 
From: [hemant@mnkcg.com](mailto:hemant@mnkcg.com) <[hemant@mnkcg.com](mailto:hemant@mnkcg.com)> 
Sent: Thursday, June 17, 2021 8:24 AM
To: [hemant@mnkcg.com](mailto:hemant@mnkcg.com); Mihai Budiu <[mbudiu@vmware.com](mailto:mbudiu@vmware.com)>; [Gergely.Pongracz@ericsson.com](mailto:Gergely.Pongracz@ericsson.com); [jnfoster@cs.cornell.edu](mailto:jnfoster@cs.cornell.edu)
Cc: [p4-design@lists.p4.org](mailto:p4-design@lists.p4.org)
Subject: RE: [P4-design] Re: new language features for "5G" and "edge" use cases
 
To store packets, I would need an array of Packet_in in P4 which is a new construct for P4.
 
Hemant

**Comment by hesingh** [comment]
P4 terminating nodes are growing.
@mihaibudiu Do you think the P4 spec could define the extern for dynamic array for use with Packet_in?

**Comment by vgurevich** [comment]
@hesingh -- as of today, it is a known fact that P4 has not been designed to work with packet payloads. The only thing it has so far, are a couple of externs, packet_in and packet_out that have fairly limited functionality.

Nevertheless, before we start adding new constructs, it would be helpful to describe in detail what you are trying to achieve. For example, as of today the packet buffering functionality is not expressable in P4 and therefore even very complex buffering engines are exposed as fixed-function components interfacing with the rest of the P4 program using intrinsic metadata. Only after you demonstrate that the current mechanisms (externs and intrinsic metadata) are not sufficient to express what you need, then we should start looking at new constructs.

So, let's go through this exercise first: describe what kind of functionality in your target you need to describe in P4, why exiting mechanisms will not work and then show how you would like the new mechanisms to look like and how it can help.

**Comment by hesingh** [comment]
My ask has been discussed in p4-design in detail after this presentation: https://opennetworking.org/wp-content/uploads/2021/05/Gergely-Pongracz-Slides.pdf

See slide 4 and the service is on a terminating node. The slide shows buffering is simple - delete/save pkt in buffer.

**Comment by hesingh** [comment]
The only issue with an extern is one vendor supports it but another may not. This is another reason to look into the issue for any other thoughts. I can think of a dynamic array in a host P4 core implementation.

**Comment by hesingh** [comment]
> @hesingh -- as of today, it is a known fact that P4 has not been designed to work with packet payloads. The only thing it has so far, are a couple of externs, packet_in and packet_out that have fairly limited functionality.
> 
> Nevertheless, before we start adding new constructs, it would be helpful to describe in detail what you are trying to achieve. For example, as of today the packet buffering functionality is not expressable in P4 and therefore even very complex buffering engines are exposed as fixed-function components interfacing with the rest of the P4 program using intrinsic metadata. Only after you demonstrate that the current mechanisms (externs and intrinsic metadata) are not sufficient to express what you need, then we should start looking at new constructs.
> 
> So, let's go through this exercise first: describe what kind of functionality in your target you need to describe in P4, why exiting mechanisms will not work and then show how you would like the new mechanisms to look like and how it can help.

See

https://groups.google.com/a/lists.p4.org/g/p4-design/c/Bk36zDYY1Uw/m/nLDNwUhjAAAJ

**Comment by vgurevich** [comment]
@hesingh -- I looked at both the presentation and the discussion. I still believe that it does not provide proper justification for a feature. 

Again, you need to start with the details of the specific target. For example, I fail to see how the functionality described on the slide number 4 cannot be implemented using a fixed-function device controlled by relatively simple intrinsic metadata, such as:

```
enum bit<8> special_buffer_cmd_t {
    ADD = 0,
    DEL = 1
}

struct intrinsic_metadata_for_the_special_buffer_t {
     special_buffer_cmd_t  cmd;
     bit<32>               timeout_us;
     bit<32>               pkt_tag;           /* Optional, but might help to identify a specific packet among many */
}
```

In addition, one can add intrinsic metadata for the parser that will indicate that the packet is being resubmitted by the buffering engine rather than being received from the wire. 

Similarly, on the slide 5 it is stated that "Buffering is gap #1", whereas there are at least two practical architectures with pretty complicated buffering that is fully supported. Again, unless your goal is to make the actual buffering algorithms programmable, there is very little that you need to have. To a large extent, by the way, buffering and queueing algorithms can be easily expressed in P4, since they are nothing more than a series of lookups, mappings, threshold comparisons and state updates (scheduling/shaping is a very different story).

Internal clock-based events mentioned in the slides can also be largely implemented without any additions to the language. All that is required is for the specific architecture to define when a specific control is going to be called and with what (intrinsic) metadata. As of today, for example, the `ingress()` or `egress()` control is called in the event where a packet is received, but nothing prevents an architecture to specify a `timer()` control (or controls) that will be called every so often (and also provide the APIs to control the periodicity). In fact, the architecture can even be defined so that the `ingress()` or `egress()` control can also be called without any packet too. 

Last, but not least, nowhere in the slides it is mentioned that one needs a dynamic array.

It is somewhat mentioned in the long email chain you attached, and there it is clearly stated that the problem can be solved with an extern. I didn't go very deep (since it's a long email chain with its own threads and sub-threads), but I think that the statement is correct -- you can always define an extern that will behave like a dynamic array. Simply have a get() and set() methods that take an index.

It is very important that when presented with the new HW, we always try to use existing mechanisms first. They are very generic and very powerful. Only if there is no way around the problem, that's when we need to start developing a new solution. Imagine, if for every new IO device, people will start asking adding new instructions to the CPUs :)

The email chain also contains the "portability" argument that I find extremely weak if not outright wrong. We need to accept that as of today, P4 cannot provide any portablity between hardware targets, while SW targets are completely free to implement any architecture and any externs.

**Comment by hesingh** [comment]
> @hesingh -- I looked at both the presentation and the discussion. I still believe that it does not provide proper justification for a feature.
> 
> Again, you need to start with the details of the specific target. For example, I fail to see how the functionality described on the slide number 4 cannot be implemented using a fixed-function device controlled by relatively simple intrinsic metadata, such as:
> 
> ```
> enum bit<8> special_buffer_cmd_t {
>     ADD = 0,
>     DEL = 1
> }
> 
> struct intrinsic_metadata_for_the_special_buffer_t {
>      special_buffer_cmd_t  cmd;
>      bit<32>               timeout_us;
>      bit<32>               pkt_tag;           /* Optional, but might help to identify a specific packet among many */
> }

The packets have identical header data - how do you generate pkt_tag? Is pkt_tag a monotonically increasing number? If so, when the pkt is freed, its pkt_tag is inserted back in the pool of ids and code manages free vs. used ids. Gets complicated. 
What else do you propose for pkt_tag?

> ```
> 
> In addition, one can add intrinsic metadata for the parser that will indicate that the packet is being resubmitted by the buffering engine rather than being received from the wire.
> 
> Similarly, on the slide 5 it is stated that "Buffering is gap #1", whereas there are at least two practical architectures with pretty complicated buffering that is fully supported. Again, unless your goal is to make the actual buffering algorithms programmable, there is very little that you need to have. To a large extent, by the way, buffering and queueing algorithms can be easily expressed in P4, since they are nothing more than a series of lookups, mappings, threshold comparisons and state updates (scheduling/shaping is a very different story).
> 

I never say slide 5 or traffic shaping is my goal.

> Internal clock-based events mentioned in the slides can also be largely implemented without any additions to the language. All that is required is for the specific architecture to define when a specific control is going to be called and with what (intrinsic) metadata. As of today, for example, the `ingress()` or `egress()` control is called in the event where a packet is received, but nothing prevents an architecture to specify a `timer()` control (or controls) that will be called every so often (and also provide the APIs to control the periodicity). In fact, the architecture can even be defined so that the `ingress()` or `egress()` control can also be called without any packet too.
> 
> Last, but not least, nowhere in the slides it is mentioned that one needs a dynamic array.
> 

The p4-design discussion mentions dynamic array by Mihai.

> It is somewhat mentioned in the long email chain you attached, and there it is clearly stated that the problem can be solved with an extern. I didn't go very deep (since it's a long email chain with its own threads and sub-threads), but I think that the statement is correct -- you can always define an extern that will behave like a dynamic array. Simply have a get() and set() methods that take an index.
> 
> It is very important that when presented with the new HW, we always try to use existing mechanisms first. They are very generic and very powerful. Only if there is no way around the problem, that's when we need to start developing a new solution. Imagine, if for every new IO device, people will start asking adding new instructions to the CPUs :)
> 
> The email chain also contains the "portability" argument that I find extremely weak if not outright wrong. We need to accept that as of today, P4 cannot provide any portablity between hardware targets, while SW targets are completely free to implement any architecture and any externs.

**Comment by vgurevich** [comment]
> The packets have identical header data - how do you generate pkt_tag? Is pkt_tag a monotonically increasing number? If so, when the pkt is freed, its pkt_tag is inserted back in the pool of ids and code manages free vs. used ids. Gets complicated. What else do you propose for pkt_tag?

It all depends on the capabilities of the hardware that needs to be described. Which is why I am still trying to encourage you to provide a detailed description of the hardware for which you want to define an architecture that somehow requires dynamic arrays or whatever else. Do you have such hardware (or, at least, the SW model of it)? If not, this discussion can be tabled until you do.

To answer your specific question, the contents of packet headers does not matter. The program should be able to assign a tag for whichever packet it chooses. This can be done incrementally or tags can be retrieved from and returned to a FIFO (T2NA architecture successfully demonstrates how FIFOs can be exposed as P4 externs), or can be assigned based on the timestamp or some internal counter, etc. etc. That's not the point.

**Comment by hesingh** [comment]
The hardware is generic compute or multi-core ARM cpu.

**Comment by vgurevich** [comment]
@hesingh -- stating that your target is a generic CPU is not helpful. As you know, it is not a goal of the language to be able to express whatever a generic CPU can do.

It is much more helpful to think of the data plane program that you plan to execute of the general-purpose CPU as combined of different blocks; the algorithms executed by some of them (i.e. header manipulation, modification, etc.) can be expressed in P4 and algorithms executed by the others cannot (and should not) be expressed in P4. The examples of the latter are shapers and schedulers. Now, you need to figure out what is the functionality of the block that you are trying to describe and, hence, what are the interfaces (in the form of intrinsic metadata) that a P4 program should employ to interface with that block. That's all.

Unless you can get into that mindset, we'll continue to be going in circles.  You continue to state that you need "something", but fail to provide a clear description of that "something", e.g. in the form of a P4 code. If there are some things missing -- no problem -- that code does not need to compile. Simply put in a fragment and explain why this is the only way you can achieve the desired functionality.

**Comment by hesingh** [comment]
@vgurevich Sorry, I have not found anything in your discussion that I don't already know.

Please wait for Mihai to reply.

I already said I am not discussing traffic shaping. The algorithm is also clear in slide four for buffering. A dynamic array is the correct data struct to use.

**Comment by mihaibudiu** [comment]
Mihai has retired from his P4 position and will no longer make spend too much time on the P4 compiler or language design. His new job requires more of his time.

**Comment by jnfoster** [comment]
We discussed this issue at a recent P4 LDWG. 

There was broad agreement that adding dynamically-allocated data structures to the core language (as opposed to an extern) would be a major change that alters some of the fundamental principles that have guided P4 since the beginning. That is not to say that the community would be dead-set against such an extension, but it would be a major new version of the language. We encourage the proponents flesh out the use cases, prototype it through the extern mechanism, and then come back with a fully-formed proposal.
